compiler/src/iree/compiler/Dialect/Flow/Transforms/FormScalarDispatches.cpp

```diff
@@ -0,0 +1,279 @@
+static bool isSliceRoot(int workload, Operation *op) {
+  return !op->getParentOfType<DispatchRegionOp>() &&
+         isScalarOperation(workload, op);
+}
+
+// Form dispatch regions from slice of the operation.
+static FailureOr<DispatchRegionOp> formDispatchRegionFromSlice(
+    RewriterBase &rewriter, Operation *rootOp, ArrayRef<Operation *> slice) {
+  OpBuilder::InsertionGuard g(rewriter);
+  rewriter.setInsertionPoint(rootOp);
+  FailureOr<DispatchRegionOp> dispatchRegionOp =
+      wrapOpInDispatchRegion(rewriter, rootOp);
+  if (failed(dispatchRegionOp)) {
+    return rootOp->emitOpError("failed to form dispatch region with root op");
+  }
+  FailureOr<DispatchRegionOp> newDispatchOp =
+      movePrecedingOpsIntoDispatchRegion(rewriter, slice,
+                                         dispatchRegionOp.value());
+  if (failed(newDispatchOp)) {
+    return dispatchRegionOp.value()->emitOpError(
+        "failed to move slice into op");
+  }
+  return newDispatchOp.value();
+}
+
+void FormScalarDispatchesPass::runOnOperation() {
+  mlir::FunctionOpInterface funcOp = getOperation();
+  MLIRContext *context = &getContext();
+
+  IRRewriter rewriter(context);
+  if (failed(unCSEClonableOps(funcOp, context))) {
+    funcOp->emitOpError("failed to uncse clonable ops");
+    return signalPassFailure();
+  }
+
+  int scalarWorkloadLimit = 1;
+  // Convenient struct to hold all operations that need to be moved into a
+  // descriptor.
+  struct DispatchRegionDescriptor {
+    Operation *rootOp;
+    SmallVector<Operation *> fusedOps;
+  };
+
+  SmallVector<DispatchRegionDescriptor> dispatches;
+  llvm::DenseMap<Operation *, Operation *> opToRootMap;
+
+  // Walk the function in postorder, reverse orded ignore all operations
+  // not immediately nested within the `funcOp`.
+  funcOp.walk<WalkOrder::PostOrder, ReverseIterator>([&](Operation *op) {
+    if (op->getParentOp() != funcOp || opToRootMap.count(op)) {
+      return;
+    }
+
+    if (!isSliceRoot(scalarWorkloadLimit, op)) {
+      return;
+    }
+
+    llvm::SetVector<Operation *> fusedOpsSet =
+        computeSliceToMoveIntoDispatch(scalarWorkloadLimit, op, opToRootMap);
+    for (Operation *sliceOp : fusedOpsSet) {
+      assert(!opToRootMap.count(sliceOp) &&
+             "trying to add same op to two dispatches");
+      opToRootMap[sliceOp] = op;
+    }
+
+    // Iterate backwards within the block to get ops that dont necessarily
+    // have producer -> consumer relationship but can still be fused.
+    Block *currBlock = op->getBlock();
+    Operation *prevOp = op;
+    bool didHorizontalFusion = false;
+    while (prevOp != &currBlock->front()) {
+      prevOp = prevOp->getPrevNode();
+
+      if (opToRootMap.count(prevOp)) {
+        if (opToRootMap[prevOp] == op) {
+          // If op is already in the same dispatch, continue.
+          continue;
+        }
+        // If the operation is part of a different dispatch region, stop
+        // growing the dispatch.
+        break;
```

Comment:
this stops the scan entirely; is that because any op that was before prevOp has already been put into a dispatch region and there's no need to scan it or is it possible for an op prior to be possible to be moved into a dispatch region but just isn't yet due to some other condition?

Reply:
yeah it does. Dont want to create cycles in the dispatches...  We use program slice to get dispatches that follow SSA use-def chains. Here we are just "horizontal fusing" the slices....

Reply:
Actually you are right. Removed it.